executor/aggregate.go

```diff
@@ -977,3 +997,283 @@ func (e *groupChecker) reset() {
+	groupOffset []int
+	// groupCount is the number of groups in the current chunk
+	groupCount int
+	// curGroupID is the group obtained by the next call to getNextGroup
+	curGroupID int
+
+	// previousLastGroupKey is the groupKey of the last group of the previous chunk
+	previousLastGroupKey string
+	// firstGroupKey and lastGroupKey are used to store the groupKey of the first and last group of the current chunk
+	firstGroupKey []byte
+	lastGroupKey  []byte
+
+	// firstRowDatums and lastRowDatums store the results of the expression evaluation for the first and last rows of the current chunk in datum
+	// They are used to encode to get firstGroupKey and lastGroupKey
+	firstRowDatums []types.Datum
+	lastRowDatums  []types.Datum
+
+	// sameGroup is used to check whether two adjacent rows belong to the same group
+	sameGroup []bool
+}
+
+func newVecGroupChecker(ctx sessionctx.Context, items []expression.Expression) *vecGroupChecker {
+	sameGroup := make([]bool, 1024)
+	return &vecGroupChecker{
+		ctx:          ctx,
+		GroupByItems: items,
+		groupCount:   0,
+		curGroupID:   0,
+		sameGroup:    sameGroup,
+	}
+}
+
+// splitIntoGroups divides the chunk into more groups which the row in the same group have the same groupKey
+// the return value areSameGroup means whether the groupKey of the first group of the newly passed chunk is equal to the groupKey of the last group left before
+// TODO: Since all the group by items are only a column reference, guaranteed by building projection below aggregation, we can directly compare data in a chunk.
+func (e *vecGroupChecker) splitIntoGroups(chk *chunk.Chunk) (areSameGroup bool, err error) {
+	// The numRows can not be zero. `FetchChild` is called before `splitIntoGroups` is called.
+	// if numRows == 0, it will be returned in `fetchChild`. See `fetchChild` for more details.
+	numRows := chk.NumRows()
+
+	e.reset()
+	e.curGroupID = 0
+	if len(e.GroupByItems) == 0 {
+		e.groupOffset = append(e.groupOffset, numRows)
+		return true, nil
+	}
+
+	if cap(e.sameGroup) < numRows {
+		e.sameGroup = make([]bool, 0, numRows)
+	}
+	e.sameGroup = append(e.sameGroup, false)
+	for i := 1; i < numRows; i++ {
+		e.sameGroup = append(e.sameGroup, true)
+	}
+
+	for _, item := range e.GroupByItems {
+		err = e.checkOneGroupByItem(item, chk, numRows)
+		if err != nil {
+			return false, err
+		}
+	}
+	e.firstGroupKey, err = codec.EncodeValue(e.ctx.GetSessionVars().StmtCtx, e.firstGroupKey, e.firstRowDatums...)
+	if err != nil {
+		return false, err
+	}
+
+	e.lastGroupKey, err = codec.EncodeValue(e.ctx.GetSessionVars().StmtCtx, e.lastGroupKey, e.lastRowDatums...)
+	if err != nil {
+		return false, err
+	}
+
+	if len(e.previousLastGroupKey) == 0 {
+		areSameGroup = false
+	} else {
+		groupKey0 := e.previousLastGroupKey
+		groupKey1 := string(hack.String(e.firstGroupKey))
+		if groupKey0 == groupKey1 {
+			areSameGroup = true
+		} else {
+			areSameGroup = false
+		}
+	}
+	e.previousLastGroupKey = string(e.lastGroupKey)
```

Comment:
```suggestion
	copy(e.previousLastGroupKey, e.lastGroupKey)
```
It's better to use `[]byte` to store `previousLastGroupKey` and do memory copy here.